make: *** [Makefile:104: .assert.opt.o] Error 1

### Question

I've been trying to compile again but there seems to be a problem. 


```
g++  -O3 -pedantic -std=c++17 -g -Wall -Wextra -Wno-unused-parameter -Wno-sign-compare -Wno-address -Wno-init-list-lifetime -MD -MP -MF .assert.opt.d -c -o .assert.opt.o assert.cc
In file included from unicode.hh:8,
                 from utf8.hh:5,
                 from string.hh:7,
                 from enum.hh:4,
                 from buffer.hh:7,
                 from buffer_utils.hh:4,
                 from assert.cc:4:
ranges.hh:106:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  106 |     struct Iterator : std::iterator<std::forward_iterator_tag,
      |                            ^~~~~~~~
In file included from /usr/include/c++/12.1.0/iterator:61,
                 from array_view.hh:5,
                 from constexpr_utils.hh:8,
                 from buffer.hh:6:
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
ranges.hh:164:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  164 |     struct Iterator : std::iterator<std::forward_iterator_tag,
      |                            ^~~~~~~~
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
ranges.hh:301:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  301 |     struct Iterator : std::iterator<std::forward_iterator_tag, ValueType>
      |                            ^~~~~~~~
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
ranges.hh:398:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
  398 |     struct Iterator : std::iterator<std::forward_iterator_tag, ValueType>
      |                            ^~~~~~~~
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
In file included from face.hh:4,
                 from face_registry.hh:4,
                 from scope.hh:5,
                 from buffer.hh:12:
color.hh: In constructor ‘constexpr Kakoune::Color::Color(unsigned char, unsigned char, unsigned char, unsigned char)’:
color.hh:54:23: error: call to non-‘constexpr’ function ‘void Kakoune::Color::validate_alpha()’
   54 |         validate_alpha();
      |         ~~~~~~~~~~~~~~^~
color.hh:58:10: note: ‘void Kakoune::Color::validate_alpha()’ declared here
   58 |     void validate_alpha();
      |          ^~~~~~~~~~~~~~
In file included from parameters_parser.hh:11,
                 from highlighter.hh:13,
                 from highlighter_group.hh:6,
                 from scope.hh:6:
string_utils.hh: At global scope:
string_utils.hh:71:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
   71 |     struct Iterator : std::iterator<std::forward_iterator_tag, StringView>
      |                            ^~~~~~~~
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
parameters_parser.hh:83:28: warning: ‘template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator’ is deprecated [-Wdeprecated-declarations]
   83 |     struct iterator : std::iterator<std::forward_iterator_tag, String>
      |                            ^~~~~~~~
/usr/include/c++/12.1.0/bits/stl_iterator_base_types.h:127:34: note: declared here
  127 |     struct _GLIBCXX17_DEPRECATED iterator
      |                                  ^~~~~~~~
make: *** [Makefile:104: .assert.opt.o] Error 1
``` 

gcc, gcc-libs, and ncurses seem to be installed.

--- a/src/parameters_parser.hh
+++ b/src/parameters_parser.hh
@@ -80,8 +80,14 @@ struct ParametersParser
     // a non empty StringView value if the switch took an argument.
     Optional<StringView> get_switch(StringView name) const;
 
-    struct iterator : std::iterator<std::forward_iterator_tag, String>
+    struct iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = String;
+        using pointer = String*;
+        using reference = String&;
+        using iterator_category = std::forward_iterator_tag;
+
         iterator(const ParametersParser& parser, size_t index)
             : m_parser(parser), m_index(index) {}
 

--- a/src/ranges.hh
+++ b/src/ranges.hh
@@ -122,9 +122,14 @@ struct FilterView
 {
     using RangeIt = IteratorOf<Range>;
 
-    struct Iterator : std::iterator<std::forward_iterator_tag,
-                                    typename std::iterator_traits<RangeIt>::value_type>
+    struct Iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = typename std::iterator_traits<RangeIt>::value_type;
+        using pointer = value_type*;
+        using reference = value_type&;
+        using iterator_category = std::forward_iterator_tag;
+
         Iterator(Filter& filter, RangeIt it, RangeIt end)
             : m_it{std::move(it)}, m_end{std::move(end)}, m_filter{&filter}
         {
@@ -180,9 +185,14 @@ struct EnumerateView
 {
     using RangeIt = IteratorOf<Range>;
 
-    struct Iterator : std::iterator<std::forward_iterator_tag,
-                                    typename std::iterator_traits<RangeIt>::value_type>
+    struct Iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = typename std::iterator_traits<RangeIt>::value_type;
+        using pointer = value_type*;
+        using reference = value_type&;
+        using iterator_category = std::forward_iterator_tag;
+
         Iterator(size_t index, RangeIt it)
             : m_index{index}, m_it{std::move(it)} {}
 
@@ -317,8 +327,14 @@ struct SplitView
                                          std::pair<IteratorOf<Range>, IteratorOf<Range>>,
                                          ValueTypeParam>;
 
-    struct Iterator : std::iterator<std::forward_iterator_tag, ValueType>
+    struct Iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = ValueType;
+        using pointer = ValueType*;
+        using reference = ValueType&;
+        using iterator_category = std::forward_iterator_tag;
+
         Iterator(RangeIt pos, const RangeIt& end, Element separator, Element escaper)
          : done{pos == end}, pos{pos}, sep{pos}, end(end), separator{std::move(separator)}, escaper{std::move(escaper)}
         {
@@ -486,8 +502,14 @@ struct ConcatView
     using ValueType = typename std::common_type_t<typename std::iterator_traits<RangeIt1>::value_type,
                                                   typename std::iterator_traits<RangeIt2>::value_type>;
 
-    struct Iterator : std::iterator<std::forward_iterator_tag, ValueType>
+    struct Iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = ValueType;
+        using pointer = ValueType*;
+        using reference = ValueType&;
+        using iterator_category = std::forward_iterator_tag;
+
         static_assert(std::is_convertible<typename std::iterator_traits<RangeIt1>::value_type, ValueType>::value, "");
         static_assert(std::is_convertible<typename std::iterator_traits<RangeIt2>::value_type, ValueType>::value, "");
 

--- a/src/regex.hh
+++ b/src/regex.hh
@@ -45,8 +45,13 @@ struct MatchResults
         bool matched = false;
     };
 
-    struct iterator : std::iterator<std::bidirectional_iterator_tag, SubMatch, size_t, SubMatch*, SubMatch>
+    struct iterator
     {
+        using difference_type = size_t;
+        using value_type = SubMatch;
+        using pointer = SubMatch*;
+        using reference = SubMatch;
+        using iterator_category = std::bidirectional_iterator_tag;
         using It = typename Vector<Iterator, MemoryDomain::Regex>::const_iterator;
 
         iterator() = default;

--- a/src/string_utils.hh
+++ b/src/string_utils.hh
@@ -68,8 +68,14 @@ String expand_tabs(StringView line, ColumnCount tabstop, ColumnCount col = 0);
 
 struct WrapView
 {
-    struct Iterator : std::iterator<std::forward_iterator_tag, StringView>
+    struct Iterator
     {
+        using difference_type = ptrdiff_t;
+        using value_type = StringView;
+        using pointer = StringView*;
+        using reference = StringView&;
+        using iterator_category = std::forward_iterator_tag;
+
         Iterator(StringView text, ColumnCount max_width);
 
         Iterator& operator++();

